cpp/DEVELOPER_GUIDE.md

```diff
@@ -9,6 +9,61 @@ cuML is thread safe so its functions can be called from multiple host threads if
 
 The implementation of cuML is single threaded.
 
+## Managing state
+### Introduction
+Every ML algo needs to store some state, eg: model and its related hyper-parameters. Thus, this section lays out guidelines for managing state along the API of cuML.
+
+### Inside `libcuml++.so` aka our C++ interface
+Functions exposed via the cuML-C++ layer must be stateless. Meaning, they must accept all the required inputs, parameters and outputs in their argument list only. In other words, the stateful API should always be a wrapper around the stateless methods, NEVER the other way around. That said, internally, these stateless functions are free to create their own temporary classes, as long as they are not exposed on the interface of `libcuml++.so`.
+
+Things which are OK to be exposed on the interface:
+1. Any [POD](https://en.wikipedia.org/wiki/Passive_data_structure).
```

Comment:
Would it be useful for developers if we also mentioned `std::is_pod` (c++11) as a way to check?

Reply:
Makes sense. Updated.